totalRP3/Core/CallbackRegistry.lua

```diff
@@ -0,0 +1,346 @@
+	local RegisterCallback = "RegisterCallback";
+	local UnregisterCallback = "UnregisterCallback";
+	local UnregisterAllCallbacks = "UnregisterAllCallbacks";
+
+	local callbacks = CallbackHandler:New(registry, RegisterCallback, UnregisterCallback, UnregisterAllCallbacks);
+	callbacks.OnUsed = OnCallbackEventUsed;
+	callbacks.OnUnused = OnCallbackEventUnused;
+
+	return callbacks;
+end
+
+function TRP3_API.InitCallbackRegistryWithEvents(registry, events)
+	local callbacks = TRP3_API.InitCallbackRegistry(registry);
+	TRP3_API.AddCallbackEventTableValidator(registry, events);
+	return callbacks;
+end
+
+function TRP3_API.AddCallbackEventValidator(registry, validator)
+	local super = registry.IsEventValid or nop;
+
+	local function IsEventValid(self, event)
+		return validator(self, event) or super(self, event);
+	end
+
+	registry.IsEventValid = IsEventValid;
+end
+
+function TRP3_API.AddCallbackEventTableValidator(registry, events)
+	local function IsEventValid(self, event)
+		return events[event] ~= nil;
+	end
+
+	TRP3_API.AddCallbackEventValidator(registry, IsEventValid);
+end
+
+-- Standalone callback registries can be created through the following functions
+-- for more local convenience.
+
+local CallbackRegistry = {};
+
+function CallbackRegistry:__init()
+	self.callbacks = TRP3_API.InitCallbackRegistry(self);
+end
+
+function CallbackRegistry:TriggerEvent(event, ...)
+	assert(TRP3_API.IsEventValid(self, event), "attempted to trigger an invalid callback event");
+	self.callbacks:Fire(event, ...);
+end
+
+function TRP3_API.CreateCallbackRegistry()
+	return TRP3_API.CreateAndInitFromPrototype(CallbackRegistry);
+end
+
+function TRP3_API.CreateCallbackRegistryWithEvents(events)
+	local registry = TRP3_API.CreateCallbackRegistry();
+	TRP3_API.AddCallbackEventTableValidator(registry, events);
+	return registry;
+end
+
+-- Callbacks registered through our utilities are packed into fully bound and
+-- optimized closures with support for multiple bound parameters as well as
+-- the late-binding through string-based method names.
+--
+-- All closures strip the first parameter (event name) from the argument list
+-- to more closely match Blizzards' behavior.
+
+local CallbackClosureFactories =
+{
+	function(f) return function(_, ...) return f(...); end; end,
+	function(f, a) return function(_, ...) return f(a, ...); end; end,
+	function(f, a, b) return function(_, ...) return f(a, b, ...); end end,
+	function(f, a, b, c) return function(_, ...) return f(a, b, c, ...); end end,
+	function(f, a, b, c, d) return function(_, ...) return f(a, b, c, d, ...); end end,
+	function(f, a, b, c, d, e) return function(_, ...) return f(a, b, c, d, e, ...); end end,
+};
+
+local CallbackMethodClosureFactories =
+{
+	function(m) return function(_, o, ...) return o[m](o, ...); end; end,
+	function(m, o) return function(_, ...) return o[m](o, ...); end; end,
+	function(m, o, a) return function(_, ...) return o[m](a, ...); end end,
```

Comment:
Gonna be honest and admit I have no clue what's going on here, but every other line in this array has o as the first argument, so this one stands out as maybe a mistake?

Also inconsistent semicolons between the `end` on both arrays, first two lines on each have them, not the others.

Reply:
It's valid - added a comment to clarify things. It's technically not used but it would have added some awkward asymmetry without it.